src/plots/histogram.rs

```diff
@@ -0,0 +1,440 @@
+            error::PlotError {
+                details: "Histograms must have finite min/max values"
+            }
+        );
+        ensure!(
+            min < max || (approx_eq!(f64, min, max) && number_of_buckets == 1),
+            error::PlotError {
+                details: "Histograms max value must be larger than its min value"
+            }
+        );
+        ensure!(
+            labels.len() == number_of_buckets,
+            error::PlotError {
+                details: "Histogram must have as many labels as buckets"
+            }
+        );
+
+        Ok(Self {
+            counts: vec![0; number_of_buckets],
+            labels: Some(labels),
+            nodata_count: 0,
+            min,
+            max,
+            measurement,
+        })
+    }
+
+    pub fn add_feature_data(&mut self, data: FeatureDataRef) -> Result<()> {
+        // TODO: implement efficiently OpenCL version
+        match data {
+            FeatureDataRef::Number(number_ref) => {
+                for &value in number_ref.data() {
+                    self.handle_data_item(value);
+                }
+            }
+            FeatureDataRef::NullableNumber(number_ref) => {
+                for (&value, is_null) in number_ref.data().iter().zip(number_ref.nulls()) {
+                    self.handle_nullable_data_item(value, is_null);
+                }
+            }
+            FeatureDataRef::Decimal(decimal_ref) => {
+                for value in decimal_ref.data().iter().map(|&v| v as f64) {
+                    self.handle_data_item(value);
+                }
+            }
+            FeatureDataRef::NullableDecimal(decimal_ref) => {
+                for (value, is_null) in decimal_ref
+                    .data()
+                    .iter()
+                    .map(|&v| v as f64)
+                    .zip(decimal_ref.nulls())
+                {
+                    self.handle_nullable_data_item(value, is_null);
+                }
+            }
+            FeatureDataRef::Categorical(categorical_ref) => {
+                for value in categorical_ref.data().iter().map(|&v| v as f64) {
+                    self.handle_data_item(value);
+                }
+            }
+            FeatureDataRef::NullableCategorical(categorical_ref) => {
+                for (value, is_null) in categorical_ref
+                    .data()
+                    .iter()
+                    .map(|&v| v as f64)
+                    .zip(categorical_ref.nulls())
+                {
+                    self.handle_nullable_data_item(value, is_null);
+                }
+            }
+            _ => {
+                return error::PlotError {
+                    details: "Cannot add non-numerical data to the histogram.",
+                }
+                .fail();
+            }
+        }
+
+        Ok(())
+    }
+
+    fn handle_data_item(&mut self, value: f64) {
+        if value < self.min || value > self.max {
+            self.nodata_count += 1;
```

Comment:
why should a value be counted as no data if it's outside of the plotted range? A plot may only visualize a section of the data

Reply:
How should I count this then?

Reply:
I would dismiss the values. a histogram of a given range should only contain values in this range. (+ actual no data bucket if the user needs the info). I guess in our current usage scenario it makes no difference because the min of the histogram is always the min of all values but this doesn't always have to be.

Reply:
Well, that's the same as the current mapping histogram implementation. There are implementations that put smaller values into the first and larger values into the last buckets, it is not that clear. Is there a definition you are referring to? Nevertheless, other NO_DATA values also do not fit into the range. By now, they are at least counted.

Reply:
I would see it like this: normally you would build a histogram on the whole data range and if you are only interested in parts of the data you extract that part from the histogram. As it is implemented, however, it is not the same to build a histogram on the whole data range and then extract a part, or to build it only on the selected range. But it should be

Reply:
6c1baa2